test/pylib/cpp/item.py

```diff
@@ -0,0 +1,142 @@
+        # Report the c++ output in its own sections
+        self.add_report_section("call", "c++", output)
+
+        if failures:
+            raise CppFailureError(failures)
+
+    def repr_failure(  # type:ignore[override]
+            self, excinfo: pytest.ExceptionInfo[BaseException], **kwargs: Any) -> str | TerminalRepr | CppFailureRepr:
+        if isinstance(excinfo.value, CppFailureError):
+            return CppFailureRepr(excinfo.value.failures)
+        return pytest.Item.repr_failure(self, excinfo)
+
+    def reportinfo(self) -> tuple[Any, int, str]:
+        return self.path, 0, self.name
+
+
+class CppFile(pytest.File):
+    """
+    Represents the C++ test file with all necessary information for test execution
+    """
+    def __init__(self, *, no_parallel_run: bool = False, modes: list[str], disabled_tests: dict[str, set[str]],
+                 run_id=None, facade: CppTestFacade, arguments: Sequence[str], parameters: list[str] = None,
+                 **kwargs: Any, ) -> None:
+        super().__init__(**kwargs)
+        self.facade = facade
+        self.modes = modes
+        self.run_id = run_id
+        self.disabled_tests = disabled_tests
+        self.no_parallel_run = no_parallel_run
+        self.parameters = parameters
+        self._arguments = arguments
+
+    def collect(self) -> Iterator[CppTestFunction]:
+        project_root = find_repo(self.path)
+        suite_dir = Path(inspect.getfile(self.facade.__class__)).resolve().parent
+        for mode in self.modes:
+            test_name = self.path.stem
+            if test_name in self.disabled_tests[mode]:
+                continue
+            executable = Path(f'{project_root}/build/{mode}/test/{suite_dir.name}') / test_name
+            tests = self.facade.list_tests(str(executable), self.no_parallel_run)
+            for test_name in tests:
+                test_id = f"{test_name}.{mode}"
+                if self.run_id:
+                    test_id = f"{test_id}.{self.run_id}"
+                if self.parameters:
+                    for index, parameter in enumerate(self.parameters):
+                        yield CppTestFunction.from_parent(self, name=f'{test_id}.{index + 1}', executable=executable,
+                                                          facade=self.facade, mode=mode, test_name=test_name,
+                                                          file_name=self.path,
+                                                          arguments=[*self._arguments, parameter])
+                else:
+                    yield CppTestFunction.from_parent(self, name=test_id, executable=executable, facade=self.facade, mode=mode,
+                                                      file_name=self.path, test_name=test_name, arguments=self._arguments)
+
+
+def find_repo(path):
+    """Find repository root from the path's parents"""
+    for path in Path(path).parents:
+        # Check whether 'path/.git' exists and is a directory
```

Comment:
It is still fragile. Sometimes CONTRIBUTING.md can be in several places as well. I think there is a better way to find it out. We can stick to the root path in pytest, since now, everything in one project. I think it will be more reliable.